packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -255,110 +299,140 @@ export class ChainService implements ChainServiceInterface {
     );
 
-    assetHolders.map(async assetHolder => {
-      const obs = this.getOrAddContractObservable(assetHolder);
-      // Fetch the current contract holding, and emit as an event
+    this.channelToSubscribers.set(channelId, [
+      ...(this.channelToSubscribers.get(channelId) ?? []),
+      subscriber,
+    ]);
+
+    assetHolders.map(assetHolder => {
+      this.setUpAssetHolderListener(assetHolder);
       const contract = this.getOrAddContractMapping(assetHolder);
       if (!contract) throw new Error('The addressToContract mapping should contain the contract');
-      const currentHolding = from(this.getInitialHoldings(contract, channelId));
-
-      const subscription = concat<ContractEvent>(
-        currentHolding,
-        obs.pipe(filter(event => event.channelId === channelId))
-      ).subscribe({
-        next: async event => {
-          switch (event.type) {
-            case Deposited:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed Deposited event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.holdingUpdated(event);
-              break;
-            case AssetTransferred:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed AssetTransferred event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.assetTransferred(event);
-              break;
-            default:
-              throw new Error('Unexpected event from contract observable');
-          }
-        },
-      });
-      const subscriptions = this.channelToSubscription.get(channelId) ?? [];
-      this.channelToSubscription.set(channelId, [...subscriptions, subscription]);
+      // Fetch the current contract holding, and emit as an event
+      this.getInitialHoldings(contract, channelId).then(initialHoldings =>
+        subscriber.holdingUpdated(initialHoldings)
+      );
     });
   }
 
-  /** Implementation note:
-   *  The following is a simplified API that assumes a single registerChannel call per channel.
-   *  If we would like to allow for multiple registrations per channel, registerChannel should return a registration ID.
-   *  unregisterChannel will require the registration ID as a parameter.
-   */
   unregisterChannel(channelId: Bytes32): void {
-    const subscriptions = this.channelToSubscription.get(channelId);
-    if (subscriptions?.length !== 1) {
-      throw new Error(
-        'Unregister channel implementation only works when there is one subscriber per channel'
-      );
+    this.channelToSubscribers.set(channelId, []);
+  }
+
+  private async whenFinalized(channelId: Bytes32, finalizesAtS: number, tryNum: number) {
+    if (tryNum > 4) return;
```

Comment:
Does this ever get used? It seem like we're not incrementing the `tryNum`?

Reply:
That code has been removed in https://github.com/statechannels/statechannels/pull/3038/commits/0d194d4b92ca7a3fbc4a52245b43f5ab19101009